tests/components/icloud/test_config_flow.py

```diff
@@ -0,0 +1,334 @@
+"""Tests for the iCloud config flow."""
+import logging
+from unittest.mock import patch, MagicMock, PropertyMock
+import pytest
+
+from pyicloud import PyiCloudService
+
+# from pyicloud.exceptions import PyiCloudFailedLoginException
+
+from homeassistant import data_entry_flow
+from homeassistant.components.icloud import config_flow
+
+# from homeassistant.components.icloud.config_flow import (
+#     CONF_TRUSTED_DEVICE,
+#     CONF_VERIFICATION_CODE,
+# )
+from homeassistant.components.icloud.const import (
+    # DOMAIN,
+    CONF_ACCOUNT_NAME,
+    CONF_GPS_ACCURACY_THRESHOLD,
+    CONF_MAX_INTERVAL,
+    DEFAULT_GPS_ACCURACY_THRESHOLD,
+    DEFAULT_MAX_INTERVAL,
+)
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+
+# from tests.common import MockConfigEntry
+
+_LOGGER = logging.getLogger(__name__)
+
+USERNAME = "username@me.com"
+PASSWORD = "password"
+ACCOUNT_NAME = "Account name 1 2 3"
+ACCOUNT_NAME_FROM_USERNAME = None
+MAX_INTERVAL = 15
+GPS_ACCURACY_THRESHOLD = 250
+
+TRUSTED_DEVICES = [
+    {"deviceType": "SMS", "areaCode": "", "phoneNumber": "*******58", "deviceId": "1"}
+]
+
+
+def mock_icloud_service_with_cookie():
+    """Return a mocked iCloud service."""
+    service = MagicMock(spec=PyiCloudService)
+    service.return_value.authenticate.return_value = None
+    service.return_value.requires_2fa = PropertyMock(return_value=False)
+    service.return_value.trusted_devices = PropertyMock(return_value=TRUSTED_DEVICES)
+    _LOGGER.info("service")
+    _LOGGER.info(service)
+    return service
+
+
+# @pytest.fixture(name="session")
+# def mock_controller_session():
+#     """Mock a successful session."""
+#     with patch("pyicloud.base.PyiCloudSession"):
+#         yield
+
+
+# @pytest.fixture(name="service")
+# def mock_controller_service():
+#     """Mock a successful service."""
+#     with patch("pyicloud.base.PyiCloudService") as service_mock:
+#         service_mock.return_value.authenticate.return_value = None
+#         service_mock.return_value.requires_2fa = True
+#         service_mock.return_value.trusted_devices = TRUSTED_DEVICES
+#         service_mock.return_value.send_verification_code.return_value = True
+#         service_mock.return_value.validate_verification_code.return_value = True
+#         yield service_mock
+
+
+@pytest.fixture(name="service_with_cookie")
+def mock_controller_service_with_cookie():
+    """Mock a successful service while already authenticate."""
+    with patch("pyicloud.base.PyiCloudSession"):
+        with patch(
+            "pyicloud.base.PyiCloudService",
+            requires_2fa=False,
+            trusted_devices=TRUSTED_DEVICES,
+        ) as service_mock:
```

Comment:
`service_mock` is the patched class. It's not the instance of the class. The class will be called and create the class instance. When the patched class is called a new `MagicMock` instance is created that has the passed keyword arguments, above, set as attributes.

Reply:
Understood, and tests are working now :sweat_smile: 

Thanks a lot for @MartinHjelmare !